### PR TITLE
store/tikv: refine the retry mechanism of region errors

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?

Problem Summary:

TiKV adds more region errors, e.g., `ReadIndexNotReady`/`ProposalInMergingMode`/`DataIsNotReady`/`RegionNotInitialized`, TiDB should handle them properly.

### What is changed and how it works?

What's Changed:

* Add the `regionScheduling` backoff configuration for backoff due to region scheduling which is separated from the `regionMiss`. The aim is to reduce backoff sleep time.
* Handle more region errors.
* Change the order of handling region errors.

### Check List 

Tests 
- No code

Side effects

### Release note 

- No release note.

### CODE REVIEW

File: store/tikv/retry/config.go [hunk 1]
```diff
@@ -89,6 +89,7 @@ var (
 	BoPDRPC      = NewConfig("pdRPC", &metrics.BackoffHistogramPD, NewBackoffFnCfg(500, 3000, EqualJitter), tikverr.NewErrPDServerTimeout(""))
 	// change base time to 2ms, because it may recover soon.
 	BoRegionMiss        = NewConfig("regionMiss", &metrics.BackoffHistogramRegionMiss, NewBackoffFnCfg(2, 500, NoJitter), tikverr.ErrRegionUnavailable)
+	BoRegionScheduling  = NewConfig("regionScheduling", &metrics.BackoffHistogramRegionScheduling, NewBackoffFnCfg(2, 500, NoJitter), tikverr.ErrRegionUnavailable)
```
Comment (sticnarf) [inline store/tikv/retry/config.go:92]: > Add the `regionScheduling` backoff configuration for backoff due to region scheduling which is separated from the `regionMiss`. The aim is to reduce backoff sleep time.

`BoRegionScheduling` config is the same as `BoRegionMiss`?

File: store/tikv/region_request.go [hunk 2]
```diff
@@ -915,17 +918,22 @@ func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, req
...
+		if seed != nil {
+			logutil.BgLogger().Debug("tikv reports `RegionNotFound` in follow-reader",
+				zap.Stringer("ctx", ctx), zap.Uint32("seed", *seed))
+			*seed = *seed + 1
+		}
+		s.regionCache.InvalidateCachedRegion(ctx.Region)
+		return false, nil
+	}
+
+	if regionErr.GetKeyNotInRegion() != nil {
```
Comment (sticnarf) [inline store/tikv/region_request.go:932]: add a log here?

File: store/tikv/region_request.go [hunk 3]
```diff
@@ -939,6 +947,7 @@ func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, req
 		}
 		return retry, errors.Trace(err)
 	}
+
 	if regionErr.GetServerIsBusy() != nil {
 		logutil.BgLogger().Warn("tikv reports `ServerIsBusy` retry later",
```
Comment (sticnarf) [inline store/tikv/region_request.go:952]: Should the log level be adjusted like others?

File: store/tikv/retry/config.go (hunk 1 above)
Comment (youjiali1995) [reply store/tikv/retry/config.go:92]: Yes. It reduces the backoff sleep time by dividing the backoff times to different configs.

File: store/tikv/region_request.go (hunk 3 above)
Comment (youjiali1995) [reply store/tikv/region_request.go:952]: I think it's qualified...

File: store/tikv/retry/config.go (hunk 1 above)
Comment (sticnarf) [reply store/tikv/retry/config.go:92]: Ah, I see...

**Review by sticnarf — APPROVED** [review]

**Review by lysu — APPROVED** [review]

**Pull request merged**